src/app/reporting/SynchronizedReportSchedulerImpl.cpp

```diff
@@ -0,0 +1,178 @@
+    mTimerDelegate->CancelTimer(this);
+}
+
+void SynchronizedReportSchedulerImpl::UnregisterReadHandler(ReadHandler * aReadHandler)
+{
+    // Verify list is populated and handler is not null
+    VerifyOrReturn((!mReadHandlerList.Empty() || (nullptr == aReadHandler)));
+
+    ReadHandlerNode * removeNode = FindReadHandlerNode(aReadHandler);
+    // Nothing to remove if the handler is not found in the list
+    VerifyOrReturn(nullptr != removeNode);
+
+    mReadHandlerList.Remove(removeNode);
+    mNodesPool.ReleaseObject(removeNode);
+
+    if (mReadHandlerList.Empty())
+    {
+        // Only cancel the timer if there are no more handlers registered
+        CancelReport(aReadHandler);
+    }
+}
+
+/// @brief Checks if the timer is active for the given ReadHandler. Since all read handlers are scheduled on the same timer, we
+/// check if the node is in the list and if the timer is active for the ReportScheduler
+bool SynchronizedReportSchedulerImpl::IsReportScheduled(ReadHandler * aReadHandler)
+{
+    ReadHandlerNode * node = FindReadHandlerNode(aReadHandler);
+    VerifyOrReturnValue(nullptr != node, false);
+    return mTimerDelegate->IsTimerActive(this);
+}
+
+/// @brief Find the smallest maximum interval possible and set it as the common maximum
+/// @return NO_ERROR if the smallest maximum interval was found, error otherwise, INVALID LIST LENGTH if the list is empty
+CHIP_ERROR SynchronizedReportSchedulerImpl::FindNextMaxInterval()
+{
+    VerifyOrReturnError(!mReadHandlerList.Empty(), CHIP_ERROR_INVALID_LIST_LENGTH);
+    System::Clock::Timestamp now      = mTimerDelegate->GetCurrentMonotonicTimestamp();
+    System::Clock::Timestamp earliest = now + Seconds16(kSubscriptionMaxIntervalPublisherLimit);
+
+    for (auto & iter : mReadHandlerList)
+    {
+        if (iter.GetMaxTimestamp() < earliest && iter.GetMaxTimestamp() > now)
+        {
+            earliest = iter.GetMaxTimestamp();
+        }
+    }
+
+    mNextMaxTimestamp = earliest;
+
+    return CHIP_NO_ERROR;
+}
+
+/// @brief Find the highest minimum timestamp possible that still respects the lowest max timestamp and sets it as the common
+/// minimum. If the max timestamp has not been updated and is in the past, or if no min timestamp is lower than the current max
+/// timestamp, this will set now as the common minimum timestamp, thus allowing the report to be sent immediately.
+/// @return NO_ERROR if the highest minimum timestamp was found, error otherwise, INVALID LIST LENGTH if the list is empty
+CHIP_ERROR SynchronizedReportSchedulerImpl::FindNextMinInterval()
+{
+    VerifyOrReturnError(!mReadHandlerList.Empty(), CHIP_ERROR_INVALID_LIST_LENGTH);
+    System::Clock::Timestamp latest = mTimerDelegate->GetCurrentMonotonicTimestamp();
+
+    for (auto & iter : mReadHandlerList)
+    {
+        if (iter.GetMinTimestamp() > latest)
+        {
+            // We do not want the new min to be set above the max for any handler
```

Comment:
Followup, I guess, for the following concern:

Let's say I have two subscriptions set up, both with 0 as min and 7200 as max.  They are ticking along, everything is fine.  Attribute changes and urgent events are reported quickly.

Now a new subscription is created with 3600 as min and 7200 as max, because it does not really care about fast updates.  Maybe it's subscribing to just events, and only non-urgent ones, so it figures it's going to mostly get reports at its max interval anyway, except when the event buffer fills.

But now the other two subscriptions will stop getting their urgent events until an hour has gone by.

I understand that we don't want to schedule multiple wake-ups for the min-intervals of multiple subscriptions, ideally, so we want to coalesce things.  But I think in the situation I described above the right behavior would be to set the timeout to something small-ish that would cover both of the "fast" subscriptions if one or both of them is reportable or something.

More generally:

1. I would think that only subscriptions that are currently reportable should probably contribute to the min-interval computation here.  A subscription that is not reportable is not blocking on its min-interval anyway, right?  If nothing is reportable, we might as well set `mNextMinTimestamp` to `mNextMaxTimestamp`.  This would address the "Find out if any handler is reportable now" bit from `CalculateNextReportTimeout`.
2. We really want to set the min-interval to the _min_ of those min-intervals for best responsiveness.
3. But that can cause too many wakeups, so we drag it out some.  But we should not drag it out "too much", whatever that means; sorting this out is what needs the followup.

Reply:
Like in the case mentioned above, checking IsReportableNow would end in all the handlers currently scheduled for a report to contribute. A solution would be to check on reportability of the handlers when the ReportTimerCallback is called and create a flag to mark them as Scheduled, and then we could check on that condition to determine if they can contribute to the mNextMinTimestamp.

That being said, you are correct with point 2 saying that for responsiveness, we would be better of getting rid of the common Min altogether, but then any attribute change would be followed by a series of trailing desync reports.

A simple solution would be to enforce a low enough max value for min interval (0 was flat out refused but maybe x seconds would be acceptable?).

Reply:
> checking IsReportableNow would end in all the handlers currently scheduled for a report to contribute. 

Right, as opposed to _all_ the handlers contributing.  That is, it doesn't matter what the min intervals of handlers that are not scheduled for a report are, since those are not blocked on min-interval, right?  So we should only consider reportable ones for that.  That's my point 1.

Points 2 and 3 we should think about, but as I said above that's best done in a followup, since that part will be a bit hard to figure out.  I think we can do better than just "cap the min interval".

Reply:
My main concern is about report currently scheduled with a min timestamp of 0. Means if handler1 and handler2 are scheduled for the same report by their max (as it should be), when handler1's report is emitted and it recalculates, it will see handler2 being reportable, and then proceed to schedule the next wake up timeout to 0 and sync its timeout with it.

Handler2 will then see handler1 being reportable on completion, and also set the timeout to 0 and sync with it, and so on. 

Since we don't want that, my idea is to add a flag that would be set in `ReportTimerCallback` that would indicate if a handler will generate a report during the current ongoing run.

Reply:
I somehow misunderstood the concern (was thinking about the issue below). I'll apply the IsReportableCheck instead.

Also opened an issue here : https://github.com/project-chip/connectedhomeip/issues/27997